### ISSUE TITLE
Slow fish completions

### ISSUE THREAD

**Issue by Kevin-Mok** [TailorDev/Watson#250]
## Context (Environment)
- Terminal: rxvt-unicode-pixbuf 9.22-2
- Shell: fish-3.0.0-1
- Version: watson 1.6.0-1
- Completion: latest version from https://github.com/TailorDev/Watson/commit/f1a2c94c1878a6f0cace8bdd509628990e319c65

## Expected Behavior
Completions as fast as `zsh`. Here I am using `watson add` (`wa` is my alias) in `zsh`:
<img src="https://user-images.githubusercontent.com/25857736/51909490-c6472d80-239a-11e9-8bba-73e63000ef1b.gif" height="150"/>
The projects almost come up instantly, and the `+` for the tags right after comes almost immediately.

## Current Behavior
Pretty unexpectedly slow completions in `fish`.
<img src="https://user-images.githubusercontent.com/25857736/51907818-fcce7980-2395-11e9-9d32-ea166e220d04.gif" height="150"/>
The projects take a few seconds to come up, and the `+` for the tags right takes bizarrely long to show up, even though it's not actually displaying any tags yet.

## Proposal
Could the speed of this be improved in any way, or is it just a problem on my own machine?

**Comment by jmaupetit** [comment]
Hi @Kevin-Mok ! Thanks for the report. We will need relevant inputs from Fish users (which I am not).

**Comment by andrewmeyer** [comment]
-- wanted to drop an update on this:

I did notice that zsh caches the output of `watson tags`, `watson projects`, and `watson frames`, where fish (from what I can tell) does not have such a feature.
in watson.fish we collect tags, frames, and projects by running the respective watson command, but `watson edit` does not appear to have the same issues that `watson start` does.  

I'm still looking at ways of speeding this up.  It is possible that the completion is being slowed down by iterating over the command looking for projects, and then iterating over it again looking for tags.

I'll keep digging

**Comment by jmaupetit** [comment]
Thanks @andrewmeyer!

**Comment by Kevin-Mok** [comment]
Yeah, so I think that the actual `watson projects` and `watson tags` commands are the bottleneck for the completion. Is `watson` scanning all the frames on-the-fly for the names?

**Comment by jmaupetit** [comment]
> Is watson scanning all the frames on-the-fly for the names?

Yes.

**Comment by andrewmeyer** [comment]
It might be worthwhile to explore adding a limit (or an option to limit) the depth watson scans for projects and tags.  In the case of this completion, There is a high probability that we would only be interested in the last 7-30 days worth of frames  (or the last 100).

I was also thinking last night about the feasibility of using memcached for either the completion side or for watson if it is installed on the system.

**Comment by jmaupetit** [comment]
AFAIR @k4nar battle tested JSON blob (frames) loading performance and we were rather surprised by the `json` module parsing efficiency. Thus, we decided to stick to a simple JSON file-based storage. And I have to say that it works pretty well (even after many years using it!). But maybe it's time to re-think this component to be more efficient.

I've not a clear idea yet about what could be the alternatives we should consider to store/request the frames database, but I have some requirements in mind to keep the original simplicity of Watson:

1. it should be a lightweight dependency (not sure Pandas is for example),
2. it should not imply running a server or a daemon,
3. it should not be cache-based, as IMO it would cause more troubles than the problem it will solve.

Of course, we can discuss those requirements. Feel free to suggest alternatives to benchmark! :heart:

**Comment by andrewmeyer** [comment]
I don't think I have enough frames to stress-test (42 frames).  I am only seeing ~200ms responses for `watson tags` `watson projects` and `watson frames`

I agree that running a separate daemon for storing the frames would be difficult requirement and should be avoided.

**Comment by jmaupetit** [comment]
We have a script to generate random fixtures [1], I think it could be modified to generate fake data to test loading/querying performances.

[1] https://github.com/TailorDev/Watson/blob/master/scripts/fuzzer.py

**Comment by jmaupetit** [comment]
> Is SQLite a viable storage alternative?

It could, but it changes many things. I would prefer keeping data unstructured.

> Also, what about the possibility of keeping a list of projects/names in a list and updating it whenever a new frame is added?

And also whenever we edit frames or rename projects, tags... This is more or less equivalent to cache invalidation :sweat_smile:

**Comment by Kevin-Mok** [comment]
So based on `watson frames | wc -l`, I'm at 56 frames now, and the problem is compounding. Here's what the completion looks like at the moment:

![watson-pre](https://user-images.githubusercontent.com/25857736/52519068-e04b0080-2c23-11e9-92fb-da8737fbfe27.gif)

So as a temporary workaround, I just created a manual cache with `watson projects > projects`, changed [this line](https://github.com/TailorDev/Watson/blob/master/watson.fish#L23) to `cat /home/kevin/.config/fish/projects` and did the same with `tags`. Here's the result:

![watson-post](https://user-images.githubusercontent.com/25857736/52519091-3b7cf300-2c24-11e9-8a68-d8e8008df50b.gif)

:sweat_smile:

**Comment by andrewmeyer** [comment]
I wrote a much larger fuzzer to test your PR with, I'm hoping it will finish adding frames in an hour or two (I might have made it too large, hah)
